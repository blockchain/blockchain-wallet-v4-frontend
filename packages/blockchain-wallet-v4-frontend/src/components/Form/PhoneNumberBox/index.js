import React from 'react'
import styled from 'styled-components'
import { prop, toLower } from 'ramda'
import PropTypes from 'prop-types'

import IntlTelInput from 'react-intl-tel-input'
import 'react-intl-tel-input/dist/libphonenumber.js'
import 'react-intl-tel-input/dist/main.css'
import { Text } from 'blockchain-info-components'
import { Remote } from 'blockchain-wallet-v4'

const Container = styled.div`
  position: relative;
  > div {
    width: 100%;
  }
  .iti-arrow {
    right: -8px;
  }
  input {
    width: 100%;
    height: 40px;
    font-size: 14px;
    ::-webkit-input-placeholder {
      opacity: 0.35;
    }
    ::-moz-placeholder {
      opacity: 0.35;
    }
  }
  * {
    outline: none;
    color: ${props => props.theme['gray-5']};
    font-family: 'Montserrat', sans-serif;
    font-weight: 300;
  }
`
const Error = styled(Text)`
  position: absolute;
  display: block;
  height: 15px;
  top: ${props => (props.errorBottom ? '40px' : '-20px')};
  right: 0;
`

class PhoneNumberBox extends React.Component {
  componentDidMount () {
    if (this.tel) {
      this.tel.onfocus = this.props.input.onFocus
    }
  }

  changeHandler = (status, value, countryData, number, id) => {
    this.props.input.onChange(number)
  }

  blurHandler = (status, value, countryData, number, id) => {
    this.props.input.onBlur(number)
  }

  bindTel = ref => {
    const tel = prop('tel', ref)
    if (tel) this.tel = tel
  }

  render () {
    const field = this.props
    const { input, defaultValue, meta, errorBottom, disabled } = field
    const { touched, error, warning } = meta
    const countryCodeField = prop('countryCode', field)
    const upperCountryCode = Remote.is(countryCodeField)
      ? countryCodeField.getOrElse('US')
      : countryCodeField || 'US'
    const countryCode = upperCountryCode && toLower(upperCountryCode)

    return (
      <Container>
        <IntlTelInput
          ref={this.bindTel}
          disabled={disabled}
          defaultValue={defaultValue || input.value}
          onPhoneNumberChange={this.changeHandler}
          onPhoneNumberBlur={this.blurHandler}
          format
          defaultCountry={countryCode}
          preferredCountries={['us', 'gb']}
          css={['intl-tel-input', 'form-control']}
          utilsScript={'libphonenumber.js'}
          placeholder='555-555-5555'
        />
        {touched &&
          error && (
            <Error
              size='12px'
              weight={300}
              color='error'
              errorBottom={errorBottom}
            >
              {error}
            </Error>
          )}
        {touched &&
          !error &&
          warning && (
            <Error
              size='12px'
              weight={300}
              color='sent'
              errorBottom={errorBottom}
            >
              {warning}
            </Error>
          )}
      </Container>
    )
  }
}

PhoneNumberBox.propTypes = {
  countryCode: PropTypes.object.isRequired
}

export default PhoneNumberBox
