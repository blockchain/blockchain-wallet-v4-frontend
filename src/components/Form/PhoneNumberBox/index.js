import React from 'react'
import styled from 'styled-components'
import { contains } from 'ramda'

import { TextInput, Text } from 'blockchain-info-components'

const Container = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: flex-start;
  width: 100%;
  height: 40px;
`
const Error = styled(Text)`
  position: absolute;
  display: block;
  top: -18px;
  right: 0;
  height: 15px;
`
const getErrorState = (meta) => {
  return !meta.touched ? 'initial' : (meta.invalid ? 'invalid' : 'valid')
}

const authorizedKeys = [' ', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+']
const handleKeyPress = event => { if (!contains(event.key, authorizedKeys)) { event.preventDefault() } }

const PhoneNumberBox = (field) => {
  const errorState = getErrorState(field.meta)

  return (
    <Container>
      <TextInput {...field.input} errorState={errorState} placeholder={field.placeholder} onKeyPress={handleKeyPress} />
      {field.meta.touched && field.meta.error && <Error size='13px' weight={300} color='error'>{field.meta.error}</Error>}
    </Container>
  )
}

export default PhoneNumberBox
