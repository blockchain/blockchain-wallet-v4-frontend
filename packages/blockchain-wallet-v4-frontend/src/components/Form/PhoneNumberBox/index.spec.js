import React from 'react'
import { shallow } from 'enzyme'
import 'jest-styled-components'
import toJson from 'enzyme-to-json'
import PhoneNumberBox from './index.js'

jest.mock('react-intl-tel-input/dist/main.css', () => jest.fn())
jest.mock('react-intl-tel-input/dist/libphonenumber.js', () => jest.fn())
jest.mock('react-intl-tel-input', () => jest.fn())

describe('PhoneNumberBox', () => {
  it('renders correctly', () => {
    const component = shallow(<PhoneNumberBox countryCode={'US'} />)
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
