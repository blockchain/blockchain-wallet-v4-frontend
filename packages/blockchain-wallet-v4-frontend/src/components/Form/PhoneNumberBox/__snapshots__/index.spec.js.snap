// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneNumberBox renders correctly 1`] = `
<PhoneNumberBox__Container>
  <Styled(mockConstructor)
    _css={
      Array [
        "intl-tel-input",
        "form-control",
      ]
    }
    defaultCountry="us"
    defaultValue=""
    format={true}
    onPhoneNumberBlur={[Function]}
    onPhoneNumberChange={[Function]}
    placeholder="555-555-5555"
    preferredCountries={
      Array [
        "us",
        "gb",
      ]
    }
    utilsScript="libphonenumber.js"
  />
</PhoneNumberBox__Container>
`;
